File: services/identity/permissions.go
package main

type Role string
type PermissionAction string
type TargetEnvironment string

const (
	RoleOwner     Role = "owner"
	RoleAdmin     Role = "admin"
	RoleDeveloper Role = "developer"
	RoleViewer    Role = "viewer"

	ActionDeploy          PermissionAction = "deploy:create"
	ActionRollback        PermissionAction = "deploy:rollback"
	ActionRepoDelete      PermissionAction = "repo:delete"
	ActionPRReview        PermissionAction = "pr:review"
	ActionReadLogs        PermissionAction = "logs:read"

	EnvProduction TargetEnvironment = "production"
	EnvPreview    TargetEnvironment = "preview"
)

type AuthContext struct {
	UserID string `json:"userId"`
	OrgID  string `json:"orgId"`
	Role   Role   `json:"role"`
}

type PermissionResult struct {
	Allowed bool   `json:"allowed"`
	Reason  string `json:"reason"`
}

// EvaluatePermission computes authorization decisions in O(1) time using role matrices and ABAC rules.
func EvaluatePermission(ctx AuthContext, action PermissionAction, env TargetEnvironment) PermissionResult {
	if ctx.Role == RoleOwner || ctx.Role == RoleAdmin {
		return PermissionResult{Allowed: true, Reason: "Full administrative permission granted."}
	}

	if ctx.Role == RoleDeveloper {
		if (action == ActionRollback || action == ActionRepoDelete) && env == EnvProduction {
			return PermissionResult{
				Allowed: false,
				Reason:  "Developer role is restricted from executing production rollbacks or repository deletions.",
			}
		}
		return PermissionResult{Allowed: true, Reason: "Developer permission granted for target environment."}
	}

	if ctx.Role == RoleViewer {
		if action == ActionPRReview || action == ActionReadLogs {
			return PermissionResult{Allowed: true, Reason: "Read-only viewer permission granted."}
		}
		return PermissionResult{Allowed: false, Reason: "Viewer role is restricted to read-only actions."}
	}

	return PermissionResult{Allowed: false, Reason: "Unknown role or unmapped action."}
}

File: services/identity/domain/permissions.ts
import { AuthContext, MemberRole, PermissionAction, PermissionResult } from '../../../packages/types/identity';

/**
 * Role Permission Matrix defining allowed actions per organization role.
 */
const ROLE_PERMISSIONS: Record<MemberRole, Set<PermissionAction>> = {
  owner: new Set([
    'repo:create',
    'repo:delete',
    'repo:push',
    'pr:review',
    'pr:merge',
    'deploy:trigger',
    'deploy:rollback',
    'env:manage',
    'org:manage',
    'billing:access',
  ]),
  admin: new Set([
    'repo:create',
    'repo:push',
    'pr:review',
    'pr:merge',
    'deploy:trigger',
    'deploy:rollback',
    'env:manage',
    'org:manage',
  ]),
  developer: new Set([
    'repo:create',
    'repo:push',
    'pr:review',
    'pr:merge',
    'deploy:trigger',
    'deploy:rollback',
  ]),
  viewer: new Set([
    'pr:review',
  ]),
};

/**
 * Pure, deterministic Fine-Grained Access Control (FGAC) Policy Evaluator.
 * 
 * Performs O(1) role permission lookups and attribute-based checks.
 */
export function evaluatePermission(
  context: AuthContext,
  action: PermissionAction,
  targetEnv: 'development' | 'preview' | 'production' = 'development'
): PermissionResult {
  const { session } = context;

  // 1. Check Session Expiration
  if (new Date(session.expiresAt).getTime() <= Date.now()) {
    return { allowed: false, reason: 'Session expired' };
  }

  // 2. Base Role Permission Check
  const roleActions = ROLE_PERMISSIONS[session.role];
  if (!roleActions || !roleActions.has(action)) {
    return {
      allowed: false,
      reason: `Role '${session.role}' is not granted action '${action}'`,
    };
  }

  // 3. Attribute-Based Environment Boundaries (ABAC)
  // Only Admin or Owner can rollback or delete in Production
  if (targetEnv === 'production') {
    if ((action === 'deploy:rollback' || action === 'repo:delete') && session.role === 'developer') {
      return {
        allowed: false,
        reason: `Action '${action}' in production environment requires Admin or Owner role`,
      };
    }
  }

  return { allowed: true };
}
